crates/client/settlement/src/sync_state.rs

```diff
@@ -0,0 +1,266 @@
+    ///
+    /// This is an external loop that is responsible for handling temporary (recoverable) errors.
+    pub async fn sync_state(
+        substrate_client: Arc<SC>,
+        settlement_provider: Box<dyn SettlementProvider<B>>,
+        retry_strategy: Box<dyn RetryStrategy<B>>,
+        starknet_spec: StarknetSpec,
+    ) {
+        loop {
+            match Self::sync_state_loop(&substrate_client, settlement_provider.as_ref(), &starknet_spec).await {
+                Ok(()) => {
+                    // Inner finality notifications stream is terminated
+                    return;
+                }
+                Err(err) => {
+                    log::error!("[settlement] {err}");
+                    match retry_strategy.can_retry(&err).await {
+                        true => {
+                            log::info!("[settlement] Retrying after error");
+                            continue;
+                        }
+                        false => panic!("Unrecoverable error in settlement thread"),
+                    }
+                }
+            }
+        }
+    }
+
+    /// This is an internal loop that listens to the new finalized blocks
+    /// and attempts to settle the state.
+    ///
+    /// It works as follows:
+    ///
+    /// 1. First of all it retrieves the latest settled state
+    /// 2. Then it starts to listen for new finality notifications
+    /// 3. For all incoming blocks with height lower than the settled one it checks the state root
+    ///    validity.
+    /// Inconsistent state root means we have a fatal error, which cannot be resolved automatically.
+    ///
+    /// 4. Once it gets up to the tip of the chain it starts to apply new state updates.
+    /// It is possible that there is a need to apply multiple state updates.
+    ///
+    /// Sync state loop operates as long as there are new blocks being finalized.
+    /// In case chain is stuck it won't update the state, even if there are pending blocks.
+    /// It is ok, since it's not a normal condition, and generally we expect that the chain will
+    /// advance indefinitely.
+    async fn sync_state_loop<SP>(
+        substrate_client: &SC,
+        settlement_provider: &SP,
+        starknet_spec: &StarknetSpec,
+    ) -> Result<(), B>
+    where
+        SP: ?Sized + SettlementProvider<B>,
+    {
+        let mut last_settled_state = settlement_provider.get_state().await?.ok_or(Error::StateNotInitialized)?;
+        log::debug!("[settlement] Last settled state {:?}", last_settled_state);
+
+        Self::verify_starknet_spec(settlement_provider, starknet_spec).await?;
+        // If we haven't reached the settled level yet (e.g. syncing from scratch) this check will pass.
+        // But we need to run it again once we are up to speed.
+        Self::verify_starknet_state(substrate_client, &last_settled_state)?;
+
+        let mut finality_notifications = substrate_client.finality_notification_stream();
```

Comment:
I don't like that we are using this in order to sync l2.
Different reason:
1) https://crates.parity.io/sc_client_api/client/trait.BlockchainEvents.html#tymethod.finality_notification_stream the doc state their is no guarantee an event is fired
2) If this thread goes down all the network node has to go done, otherwise we may miss some events

I would rather use a combination of https://crates.parity.io/sp_blockchain/trait.Backend.html#tymethod.last_finalized (in order to get the last finalized one) and https://crates.parity.io/sp_blockchain/trait.HeaderBackend.html#tymethod.header (to get every block from last_settled to last_finalized). This way this thread can be down, we can reboot the node and it will still correctly sync

Reply:
```rust
fn sync() -> !  {
let last_settled = get_last_settled();
let last_finalized = get_last_finalized();

  if last_settled.number == last_finalized.number {
     sleep.await;
     continue; 
  }
  if last_settled.number > last_finalized.number {
     panic!();
  }

  for block_number in last_settled.number..last_finalized.number {
     let block = get_block(block_number);
     <do our thing>
  }

}
```

Reply:
Regarding the second item - this is handled in the current implementation, the thread can catch up with the head.
It also handles the situations where Madara state is behind the settled one (e.g. you are bootstrapping a new node) - in your snipped you panic, which is prob not what you'd want.

Overall, listening the finality notification stream is basically the same as I would poll last_finalized() method. Given that we cannot miss events, it's probably ok even if it's not guaranteed that event is not fired, as long as there will be another one next.

Reply:
Okay, I missed that part. So you fallback yo the solution I described if the chain is not up to date.
Seems okay